### ISSUE TITLE
Queue tasks

### ISSUE THREAD

**Issue by annevk** [whatwg/fs#3]
While preparing #1 I noticed that many promises are resolved and rejected from in parallel steps. Instead this has to be done from tasks. https://html.spec.whatwg.org/#queue-a-global-task seems most appropriate here, not entirely sure if this needs a new task source. Perhaps we should have one for storage in general?

**Comment by a-sully** [comment]
I'm supportive of this!

Seeing as AccessHandles is supposed to be the fastest storage primitive ever exposed to the web, does it makes sense to provide the Access Handles API its own task source? (in addition to one for storage more broadly)

**Comment by annevk** [comment]
Yeah, I think that could be reasonable.

cc @smaug---- @domenic

**Comment by a-sully** [comment]
Revisiting this... All methods on the SyncAccessHandle are now sync (as per #7), so the argument made in my above comment about wanting a separate task source for this API is somewhat less relevant (although we'd like to eventually support an async version of SyncAccessHandles which is similarly performant - see #41).

It seems like we should have a new task source. @annevk do you have a preference for whether it's for this API or for storage in general?

**Comment by annevk** [comment]
I think ideally we consider which tasks need to happen in order across storage specifications and which can go out-of-order. And then put those in separate sources?

**Comment by a-sully** [comment]
Hmm good point. Thinking about this a bit more.... Writes using any storage API (including this one) can be rejected if the site is out of quota. So do all storage APIs need to use same task source for writes, since [they all interact with the Quota Management API](https://www.w3.org/TR/quota-api/#accessing-storagequota-interface)?

There are a couple of other [[1](https://www.w3.org/TR/IndexedDB/#request-source), [2](https://www.w3.org/TR/FileAPI/#blobreader-task-source)] storage-related task sources that I'm aware of, but they both queue operations that are read-only (i.e. don't affect quota)

**Comment by annevk** [comment]
That's a good point too.

So, the tasks are for informing the event loop about the result of a storage operation. So yes, write storage operations need to be in order (at least per site), but reporting whether that succeeded or not back to JavaScript (this involves tasks) doesn't necessarily have to be in the same order, but we might as well stick to a single task source for that I think.

**Comment by a-sully** [comment]
I was about to file an issue in whatwg/storage but stumbled upon https://github.com/whatwg/storage/issues/89. I don't personally understand much of the discussion happening on that issue - is there more to it than defining a new task source for write operations in the storage spec and then pointing all the storage API specs, including this one, to it?

**Comment by annevk** [comment]
It's better now that we have "queue a global task". We should probably define a wrapper around that called "queue a storage task" as well as the "storage task source" in the Storage Standard. Would you be willing to take that on?

**Comment by a-sully** [comment]
Filed https://github.com/whatwg/storage/pull/155 (which seems to be erroring on something which doesn't seem to be my fault... :) )

**Comment by a-sully** [comment]
This was addressed in #95. All promises are now rejected/resolved on the new storage task source (in this spec, at least. See https://github.com/WICG/file-system-access/issues/420)
